python/dgl/readout.py

```diff
@@ -5,79 +5,44 @@
 
 from .base import DGLError
 from . import backend as F
+from . import segment
 
-__all__ = ['sum_nodes', 'sum_edges', 'mean_nodes', 'mean_edges',
+__all__ = ['readout_nodes', 'readout_edges',
+           'sum_nodes', 'sum_edges', 'mean_nodes', 'mean_edges',
            'max_nodes', 'max_edges', 'softmax_nodes', 'softmax_edges',
            'broadcast_nodes', 'broadcast_edges', 'topk_nodes', 'topk_edges']
 
-READOUT_ON_ATTRS = {
-    'nodes': ('ndata', 'batch_num_nodes', 'number_of_nodes'),
-    'edges': ('edata', 'batch_num_edges', 'number_of_edges'),
-}
-
-def _sum_on(graph, typestr, feat, weight):
-    """Internal function to sum node or edge features.
-
-    Parameters
-    ----------
-    graph : DGLGraph
-        The graph.
-    typestr : str
-        'nodes' or 'edges'
-    feat : str
-        The feature field name.
-    weight : str
-        The weight field name.
+def readout_nodes(graph, feat, weight=None, *, op='sum', ntype=None):
+    """Generate a graph-level representation by aggregating node features
+    :attr:`feat`.
 
-    Returns
-    -------
-    tensor
-        The (weighted) summed node or edge features.
-    """
-    data_attr, batch_num_objs_attr, _ = READOUT_ON_ATTRS[typestr]
-    data = getattr(graph, data_attr)
-    feat = data[feat]
-
-    if weight is not None:
-        weight = data[weight]
-        weight = F.reshape(weight, (-1,) + (1,) * (F.ndim(feat) - 1))
-        feat = weight * feat
-
-    n_graphs = graph.batch_size
-    batch_num_objs = getattr(graph, batch_num_objs_attr)
-    seg_id = F.zerocopy_from_numpy(np.arange(n_graphs, dtype='int64').repeat(batch_num_objs))
-    seg_id = F.copy_to(seg_id, F.context(feat))
-    y = F.unsorted_1d_segment_sum(feat, seg_id, n_graphs, 0)
-    return y
-
-def sum_nodes(graph, feat, weight=None):
-    """Sums all the values of node field :attr:`feat` in :attr:`graph`, optionally
-    multiplies the field by a scalar node field :attr:`weight`.
+    The function is commonly used as a *readout* function on a batch of graphs
+    to generate graph-level representation. Thus, the result tensor shape
+    depends on the batch size of the input graph. Given a graph of batch size
+    :math:`B`, and a feature size of :math:`D`, the result shape will be
+    :math:`(B, D)`, with each row being the aggregated node features of each
+    graph.
 
```

Comment:
Briefly describe the behavior in the case of multiple node types and give an example?